File: kybernetes_demos/src/keep_heading.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('kybernetes_motion')
import rospy
import math

from   std_msgs.msg    import Bool
from   std_msgs.msg    import Float32

# Publisher for the enabled status
velocityPublisher = rospy.Publisher('/motion_interface/velocity', Float32)
directionPublisher = rospy.Publisher('/motion_interface/direction', Float32)

# Global data
desiredHeading = 0.0
headingFound = False
direction = 0.0

# The velocity callback
def enabled_callback(on):
	# Alert the log to the change in state
	fuck = "this"

# Normalize the degrees
def normaliseD(degrees):
	# Normalise
	deg = degrees
	if deg >= 360:
		while deg >= 360: deg -= 360
	elif deg < 0:
		while deg < 0: deg += 360
	return deg

# Get shortest distance
def getShortAngle(a, b):
	angle = abs(a - b) % 360
	if angle > 180:
		angle = 360 - angle

	if a < b: angle = angle * -1
	return angle

# The direction callback
def heading_callback(heading):
	# Alert the log to the change in velocity
	global desiredHeading
	global headingFound
	
	_heading = normaliseD(heading.data)

	if headingFound == False:
		headingFound = True
		desiredHeading = _heading

	difference = getShortAngle(_heading, desiredHeading)
	if difference > 33.3: difference = 33.3
	if difference < -33.3: difference = -33.3
	directionPublisher.publish(Float32(difference * 3))
	rospy.loginfo("Diff = %f", difference * 3)

# Main processing (routing) node - takes the ros messages and publishes them to the device
def keep_heading():
	rospy.init_node('keep_heading', anonymous = False)
	global desiredHeading
	global headingFound
	desiredHeading = 0.0
	headingFound = False

	# Subscribe to velocity and direction messages
	rospy.Subscriber("/motion_interface/enabled", Bool, enabled_callback)
	rospy.Subscriber("/imu0/state", Float32, heading_callback)
	
	# Perform a loop checking on the enabled status of the drivers
	while not rospy.is_shutdown():
		velocityPublisher.publish(Float32(80.0))
		rospy.sleep(0.25)

# Start the main class
if __name__ == '__main__':
	try:
		keep_heading()
	except rospy.ROSInterruptException: pass

File: kybernetes_demos/src/keep_heading.cpp
#include <ros/ros.h>
#include <std_msgs/Float32.h>
#include <sensor_msgs/Imu.h>
#include <kdl/frames.hpp>

#include <kybernetes_demos/pid.h>

ros::Publisher velocityPublisher, directionPublisher;
sPID           pid;

// Angle Helpers
double normaliseDegrees(double degrees)
{
    double _d = degrees;
    if(_d >= 360.0)
        while(_d >= 360.0) _d -= 360.0;
    else if(_d <= 0.0)
        while(_d <= 0.0) _d += 360.0;
    return _d;
}

double getShortAngle(double a, double b)
{
    double angle = fmod(fabs(a - b), 360.0);
    if(angle > 180.0) angle = 360.0 - angle;
    if(a < b) angle *= -1;
    return angle;
}

// IMU callback
void imu_callback(const sensor_msgs::Imu& _imu)
{
    // We need to convert back to roll, pitch, and yaw
    double roll = 0.0, pitch = 0.0, yaw = 0.0;
    KDL::Rotation::Quaternion(_imu.orientation.x, 
                              _imu.orientation.y, 
                              _imu.orientation.z, 
                              _imu.orientation.w).GetRPY(roll, 
                                                         pitch, 
                                                         yaw);
    // Do something with RPY (Feed into steering PID)
    std_msgs::Float32 _position;
    std_msgs::Float32 _velocity;
    pid.pv = getShortAngle(0.0, yaw * (180 / M_PI));
    pid_update(&pid, 0.1);
    _position.data = pid.cv;
    _velocity.data = 50.0;
    //pid_dump(&pid);
    ROS_INFO("Input: %f, Output: %f", pid.pv, pid.cv);
    directionPublisher.publish(_position);
    velocityPublisher.publish(_velocity);
}

// Main Function
int main(int argc, char** argv)
{
    // Init the ROS node system
    ros::init(argc, argv, "keep_heading");
    ros::NodeHandle n;
    pid_init(&pid);
    pid.cv_min = -100;
    pid.cv_max =  100;
    
    // Subscribers and Publishers
    ros::Subscriber imu_subscriber = n.subscribe("/imu0/state", 1, &imu_callback);
    velocityPublisher = n.advertise<std_msgs::Float32>("/motion_interface/velocity", 1000);
    directionPublisher = n.advertise<std_msgs::Float32>("/motion_interface/direction", 1000);
    
    // Loop forever
    ros::spin();
    
    // Return success
    return 0;
}
